File: src/backend/langgraph/nodes/agentNode.ts
import type { LoopContext } from "src/loopContext";
import type { Provider } from "src/services/api/types";
import type { LoopEvent } from "src/types/events";
import { driveProvider } from "../adapters/providerAdapter";
import type { AgentState } from "../state";

const MAX_ITERATIONS = 24;

export interface AgentNodeDeps {
  provider: Provider;
  model?: string;
  ctx: LoopContext;
  emit: (ev: LoopEvent) => void;
}

export const agentNode =
  ({ provider, model, ctx, emit }: AgentNodeDeps) =>
  async (state: typeof AgentState.State) => {
    if (state.iter >= MAX_ITERATIONS) {
      emit({
        type: "error",
        error: new Error(`agent loop reached max iterations (${MAX_ITERATIONS})`),
      });
      return { iter: state.iter };
    }
    emit({ type: "iteration_start", n: state.iter + 1 });
    const turn = await driveProvider({
      provider,
      messages: state.messages,
      ctx,
      ...(model !== undefined ? { model } : {}),
      signal: ctx.abortSignal,
      onEvent: (e) => {
        if (e.type === "assistant_text" && e.text) emit({ type: "assistant_text", delta: e.text });
        else if (e.type === "usage")
          emit({
            type: "usage",
            inputTokens: e.inputTokens as number,
            outputTokens: e.outputTokens as number,
          });
        else if (e.type === "error") emit({ type: "error", error: e.error as Error });
      },
    });
    const blocks: import("src/types/message").AssistantBlock[] = [];
    if (turn.text) blocks.push({ type: "text", text: turn.text });
    for (const tu of turn.toolUses) {
      let input: unknown = {};
      try {
        input = JSON.parse(tu.argsPartial || "{}");
      } catch {
        /* keep {} */
      }
      blocks.push({ type: "tool_use", id: tu.id, name: tu.name, input });
    }
    if (turn.stopReason === "max_tokens") {
      emit({ type: "error", error: new Error("provider response stopped: max tokens reached") });
    }
    return {
      messages: [{ role: "assistant", content: blocks }],
      iter: state.iter + 1,
    };
  };

File: python/src/docker_agent/backend/langgraph/nodes/agent_node.py
"""Agent node: call provider, emit events, return assistant message.

Parity: ``src/backend/langgraph/nodes/agentNode.ts``.
"""

from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from docker_agent.backend.langgraph.adapters.provider_adapter import drive_provider
from docker_agent.backend.langgraph.state import AgentState
from docker_agent.services.api.types import Provider
from docker_agent.types.events import AssistantText, Error, IterationStart, Usage
from docker_agent.types.message import AssistantBlock, AssistantMessage

MAX_ITERATIONS = 24


@dataclass
class AgentNodeDeps:
    provider: Provider
    ctx: Any
    emit: Callable[[Any], None]
    model: str | None = None


async def agent_node(deps: AgentNodeDeps, state: AgentState) -> dict[str, Any]:
    if state.iter >= MAX_ITERATIONS:
        deps.emit(
            Error(error=RuntimeError(f"agent loop reached max iterations ({MAX_ITERATIONS})"))
        )
        return {"iter": state.iter}

    deps.emit(IterationStart(n=state.iter + 1))

    def on_event(ev: Any) -> None:
        if ev.type == "assistant_text" and ev.text:
            deps.emit(AssistantText(delta=ev.text))
        elif ev.type == "usage":
            deps.emit(
                Usage(
                    inputTokens=ev.input_tokens or 0,
                    outputTokens=ev.output_tokens or 0,
                )
            )
        elif ev.type == "error":
            deps.emit(Error(error=ev.error))

    turn = await drive_provider(
        provider=deps.provider,
        messages=state.messages,
        ctx=deps.ctx,
        model=deps.model,
        on_event=on_event,
        signal=deps.ctx.abort_signal,
    )

    blocks: list[Any] = []
    if turn.text:
        blocks.append(AssistantBlock.model_validate({"type": "text", "text": turn.text}))
    for tu in turn.tool_uses:
        try:
            input_data = json.loads(tu["args_partial"] or "{}")
        except Exception:
            input_data = {}
        blocks.append(
            AssistantBlock.model_validate(
                {"type": "tool_use", "id": tu["id"], "name": tu["name"], "input": input_data}
            )
        )

    if turn.stop_reason == "max_tokens":
        deps.emit(Error(error=RuntimeError("provider response stopped: max tokens reached")))

    return {
        "messages": [AssistantMessage(content=blocks)],
        "iter": state.iter + 1,
    }